src/api/exchange/ExchangeApi.ts

```diff
@@ -101,21 +115,175 @@ export interface Order {
   remainingAmount: BN
 }
 
+/**
+ * BaseTradeEvent uses only info available on the emitted Trade event
+ */
+export interface BaseTradeEvent {
+  // order related
+  orderId: string
+  sellTokenId: number
+  buyTokenId: number
+  sellAmount: BN
+  buyAmount: BN
+  // block related
+  txHash: string
+  txIndex: number
+  blockNumber: number
+  id: string // txHash + | + txIndex
+}
+
+/**
+ * TradeEvent enriches BaseTradeEvent with block, order and token data
+ */
+export interface TradeEvent extends BaseTradeEvent {
+  batchId: number
+  hashKey: string // orderId + batchId, to find reverts
+  indexOnBatch: number // tracks trade position on batch, in case of reverts
+  time: Date
+  buyToken: TokenDetails
+  sellToken: TokenDetails
+  limitPrice: BigNumber
+  fillPrice: BigNumber
+  reverted: boolean
+}
+
 export interface GetOrdersPaginatedResult {
   orders: AuctionElement[]
   nextIndex?: number
 }
 
+// TODO: move to const/config
+const CONTRACT_DEPLOYMENT_BLOCK = {
+  1: 9340147,
+  4: 5844678,
+}
+
+// Syntactic sugar. Aliasing event types to make code cleaner
+type TradeSubscription = ReturnType<BatchExchangeContract['events']['Trade']>
+// type TradeEventData = TradeSubscription extends Subscription<infer U> ? U : never
+
+interface Subscriptions {
+  trade: { [networkId: number]: TradeSubscription }
+}
 /**
  * Basic implementation of Stable Coin Converter API
  */
 export class ExchangeApiImpl extends DepositApiImpl implements ExchangeApi {
+  private subscriptions: Subscriptions = { trade: {} }
+
   public constructor(injectedDependencies: Params) {
     super(injectedDependencies)
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     ;(window as any).exchange = this._contractPrototype
   }
 
+  public async getPastTrades({ userAddress, networkId }: PastEventsParams): Promise<BaseTradeEvent[]> {
```

Comment:
Maybe this question could be asked in other APIs, but what happens if we pass a networkId that is not the one the web3 provider is connected to?

Reply:
Contract will not be found on the network and the call will fail.
Same for all APIs I believe.

Reply:
I like how we are declarative with the parameters. But still there's sth odd in that we have the parameters, but the "internal state" also plays a role.

If we pass `networkId = 1` and web3 points to rinkeby, it'll have a contract, just the `userAddress`, so we will just get other events, or no events if the user didn't trade.

Again. I think this is weird, but is not from this PR, and maybe is not something we want to worry too much right now. Just flagging it

Reply:
Yeah, this is most likely the source for all of our "did you run out of gas?" exceptions we get when switching networks or loading the page.